### ISSUE TITLE
Zvol size more than requested? 

### ISSUE THREAD

**Issue by ishioni** [democratic-csi/democratic-csi#136]
I requested a 300MB pvc backed by iscsi on truenas scale. Lo and behold, truenas is showing the volume is taking up more than 300 of those megabytes, and the size keeps growing once i actually start writing data to it. I thought it might have been a bug with the ui, so I checked on the cli, and nope, that's actually how it is. Am I missing something here? This happens to all of my iscsi zvols, which is kinda concering. Especially for my nextcloud backing store, which is supposed to be 100G, but with data written actually takes up 170G :/ (with df -h on the mounted fs showing around 70GB of data)

This is on a RAID1 pool if it makes any difference

```
NAME                                                      PROPERTY                                               VALUE                                                                                                                                                                     SOURCE
SSD/k8s/volumes/pvc-bdd7c552-ca93-4400-9974-57d23763fd05  type                                                   volume                                                                                                                                                                    -
SSD/k8s/volumes/pvc-bdd7c552-ca93-4400-9974-57d23763fd05  creation                                               Sun Nov 21 22:56 2021                                                                                                                                                     -
SSD/k8s/volumes/pvc-bdd7c552-ca93-4400-9974-57d23763fd05  used                                                   170G                                                                                                                                                                      -
SSD/k8s/volumes/pvc-bdd7c552-ca93-4400-9974-57d23763fd05  available                                              337G                                                                                                                                                                      -
SSD/k8s/volumes/pvc-bdd7c552-ca93-4400-9974-57d23763fd05  referenced                                             68.2G                                                                                                                                                                     -
SSD/k8s/volumes/pvc-bdd7c552-ca93-4400-9974-57d23763fd05  compressratio                                          1.01x                                                                                                                                                                     -
SSD/k8s/volumes/pvc-bdd7c552-ca93-4400-9974-57d23763fd05  reservation                                            none                                                                                                                                                                      default
SSD/k8s/volumes/pvc-bdd7c552-ca93-4400-9974-57d23763fd05  volsize                                                100G                                                                                                                                                                      local
SSD/k8s/volumes/pvc-bdd7c552-ca93-4400-9974-57d23763fd05  volblocksize                                           16K                                                                                                                                                                       -
SSD/k8s/volumes/pvc-bdd7c552-ca93-4400-9974-57d23763fd05  checksum                                               on                                                                                                                                                                        default
SSD/k8s/volumes/pvc-bdd7c552-ca93-4400-9974-57d23763fd05  compression                                            zstd                                                                                                                                                                      inherited from SSD/k8s
SSD/k8s/volumes/pvc-bdd7c552-ca93-4400-9974-57d23763fd05  readonly                                               off                                                                                                                                                                       inherited from SSD
SSD/k8s/volumes/pvc-bdd7c552-ca93-4400-9974-57d23763fd05  createtxg                                              966134                                                                                                                                                                    -
SSD/k8s/volumes/pvc-bdd7c552-ca93-4400-9974-57d23763fd05  copies                                                 1                                                                                                                                                                         inherited from SSD/k8s/volumes
SSD/k8s/volumes/pvc-bdd7c552-ca93-4400-9974-57d23763fd05  refreservation                                         102G                                                                                                                                                                      local
SSD/k8s/volumes/pvc-bdd7c552-ca93-4400-9974-57d23763fd05  guid                                                   17906720968296634812                                                                                                                                                      -
SSD/k8s/volumes/pvc-bdd7c552-ca93-4400-9974-57d23763fd05  primarycache                                           all                                                                                                                                                                       default
SSD/k8s/volumes/pvc-bdd7c552-ca93-4400-9974-57d23763fd05  secondarycache                                         all                                                                                                                                                                       default
SSD/k8s/volumes/pvc-bdd7c552-ca93-4400-9974-57d23763fd05  usedbysnapshots                                        85.7M                                                                                                                                                                     -
SSD/k8s/volumes/pvc-bdd7c552-ca93-4400-9974-57d23763fd05  usedbydataset                                          68.2G                                                                                                                                                                     -
SSD/k8s/volumes/pvc-bdd7c552-ca93-4400-9974-57d23763fd05  usedbychildren                                         0B                                                                                                                                                                        -
SSD/k8s/volumes/pvc-bdd7c552-ca93-4400-9974-57d23763fd05  usedbyrefreservation                                   102G                                                                                                                                                                      -
SSD/k8s/volumes/pvc-bdd7c552-ca93-4400-9974-57d23763fd05  logbias                                                latency                                                                                                                                                                   default
SSD/k8s/volumes/pvc-bdd7c552-ca93-4400-9974-57d23763fd05  objsetid                                               39541                                                                                                                                                                     -
SSD/k8s/volumes/pvc-bdd7c552-ca93-4400-9974-57d23763fd05  dedup                                                  off                                                                                                                                                                       inherited from SSD
SSD/k8s/volumes/pvc-bdd7c552-ca93-4400-9974-57d23763fd05  mlslabel                                               none                                                                                                                                                                      default
SSD/k8s/volumes/pvc-bdd7c552-ca93-4400-9974-57d23763fd05  sync                                                   standard                                                                                                                                                                  inherited from SSD
SSD/k8s/volumes/pvc-bdd7c552-ca93-4400-9974-57d23763fd05  refcompressratio                                       1.01x                                                                                                                                                                     -
SSD/k8s/volumes/pvc-bdd7c552-ca93-4400-9974-57d23763fd05  written                                                460K                                                                                                                                                                      -
SSD/k8s/volumes/pvc-bdd7c552-ca93-4400-9974-57d23763fd05  logicalused                                            69.2G                                                                                                                                                                     -
SSD/k8s/volumes/pvc-bdd7c552-ca93-4400-9974-57d23763fd05  logicalreferenced                                      69.0G                                                                                                                                                                     -
SSD/k8s/volumes/pvc-bdd7c552-ca93-4400-9974-57d23763fd05  volmode                                                default                                                                                                                                                                   default
SSD/k8s/volumes/pvc-bdd7c552-ca93-4400-9974-57d23763fd05  snapshot_limit                                         none                                                                                                                                                                      default
SSD/k8s/volumes/pvc-bdd7c552-ca93-4400-9974-57d23763fd05  snapshot_count                                         none                                                                                                                                                                      default
SSD/k8s/volumes/pvc-bdd7c552-ca93-4400-9974-57d23763fd05  snapdev                                                hidden                                                                                                                                                                    default
SSD/k8s/volumes/pvc-bdd7c552-ca93-4400-9974-57d23763fd05  context                                                none                                                                                                                                                                      default
SSD/k8s/volumes/pvc-bdd7c552-ca93-4400-9974-57d23763fd05  fscontext                                              none                                                                                                                                                                      default
SSD/k8s/volumes/pvc-bdd7c552-ca93-4400-9974-57d23763fd05  defcontext                                             none                                                                                                                                                                      default
SSD/k8s/volumes/pvc-bdd7c552-ca93-4400-9974-57d23763fd05  rootcontext                                            none                                                                                                                                                                      default
SSD/k8s/volumes/pvc-bdd7c552-ca93-4400-9974-57d23763fd05  redundant_metadata                                     all                                                                                                                                                                       default
SSD/k8s/volumes/pvc-bdd7c552-ca93-4400-9974-57d23763fd05  encryption                                             off                                                                                                                                                                       default
SSD/k8s/volumes/pvc-bdd7c552-ca93-4400-9974-57d23763fd05  keylocation                                            none                                                                                                                                                                      default
SSD/k8s/volumes/pvc-bdd7c552-ca93-4400-9974-57d23763fd05  keyformat                                              none                                                                                                                                                                      default
SSD/k8s/volumes/pvc-bdd7c552-ca93-4400-9974-57d23763fd05  pbkdf2iters                                            0                                                                                                                                                                         default
SSD/k8s/volumes/pvc-bdd7c552-ca93-4400-9974-57d23763fd05  democratic-csi:freenas_iscsi_targettoextent_id         57                                                                                                                                                                        local
SSD/k8s/volumes/pvc-bdd7c552-ca93-4400-9974-57d23763fd05  democratic-csi:volume_context_provisioner_driver       freenas-api-iscsi                                                                                                                                                         local
SSD/k8s/volumes/pvc-bdd7c552-ca93-4400-9974-57d23763fd05  org.truenas:description                                nextcloud/nextcloud-nextcloud                                                                                                                                             local
SSD/k8s/volumes/pvc-bdd7c552-ca93-4400-9974-57d23763fd05  democratic-csi:freenas_iscsi_assets_name               pvc-bdd7c552-ca93-4400-9974-57d23763fd05                                                                                                                                  local
SSD/k8s/volumes/pvc-bdd7c552-ca93-4400-9974-57d23763fd05  democratic-csi:csi_share_volume_context                {"node_attach_driver":"iscsi","portal":"10.4.1.2:3260","portals":"","interface":"","iqn":"iqn.2005-10.org.freenas.ctl:pvc-bdd7c552-ca93-4400-9974-57d23763fd05","lun":0}  local
SSD/k8s/volumes/pvc-bdd7c552-ca93-4400-9974-57d23763fd05  democratic-csi:provision_success                       true                                                                                                                                                                      local
SSD/k8s/volumes/pvc-bdd7c552-ca93-4400-9974-57d23763fd05  democratic-csi:managed_resource                        true                                                                                                                                                                      local
SSD/k8s/volumes/pvc-bdd7c552-ca93-4400-9974-57d23763fd05  democratic-csi:freenas_iscsi_extent_id                 9                                                                                                                                                                         local
SSD/k8s/volumes/pvc-bdd7c552-ca93-4400-9974-57d23763fd05  democratic-csi:csi_volume_name                         pvc-bdd7c552-ca93-4400-9974-57d23763fd05                                                                                                                                  local
SSD/k8s/volumes/pvc-bdd7c552-ca93-4400-9974-57d23763fd05  democratic-csi:freenas_iscsi_target_id                 9                                                                                                                                                                         local
SSD/k8s/volumes/pvc-bdd7c552-ca93-4400-9974-57d23763fd05  org.truenas:managedby                                  10.4.1.8                                                                                                                                                                  local
SSD/k8s/volumes/pvc-bdd7c552-ca93-4400-9974-57d23763fd05  democratic-csi:volume_context_provisioner_instance_id  quadrant                                                                                                                                                                  local
```

**Comment by travisghansen** [comment]
Welcome! zfs is pretty complex so it can be pretty easy to get lost with some of this stuff. There are myriad of things that can contribute to sizes not matching expectations.

I would start looking at something like this: https://www.reddit.com/r/zfs/comments/d1u23o/zfs_zvol_using_way_more_data_than_volsize/

**Comment by ishioni** [comment]
@travisghansen I've actually narrowed it down - https://github.com/democratic-csi/democratic-csi/blob/a02640bec5a974013b5d423523da528956c56aa7/src/driver/freenas/api.js#L2916

It keeps setting refreservations, even though i have it disabled. Can you spot anything obviously wrong in my setup? https://github.com/ishioni/k3s-cluster/blob/master/cluster/apps/storage/democratic-csi/helm-release.yaml#L33

**Comment by travisghansen** [comment]
If you unset that the usage goes back expected levels? I’d have to do some testing but that option may not be super effective as it may get set automatically for zvols anyway.

**Comment by ishioni** [comment]
Yep, zfs set refreservations=0 gave me the result i was looking for, and goes in live with how truenas csp does it as well

**Comment by travisghansen** [comment]
Interesting. I’ll do some testing to see what the various combinations do and commit a fix as necessary.

Thanks for the research!

**Comment by travisghansen** [comment]
So this pretty much explains it all (including the behavior you're seeing):

```
volsize=size
For volumes, specifies the logical size of the volume. By default, creating a volume establishes a reservation of equal size. For storage pools with a version number of 9 or higher, a refreservation is set instead. Any changes to volsize are reflected in an equivalent change to the reservation (or refreservation). The volsize can only be set to a multiple of volblocksize, and cannot be zero.
The reservation is kept equal to the volume's logical size to prevent unexpected behavior for consumers. Without the reservation, the volume could run out of space, resulting in undefined behavior or data corruption, depending on how the volume is used. These effects can also occur when the volume size is changed while it is in use (particularly when shrinking the size). Extreme care should be used when adjusting the volume size.

Though not recommended, a "sparse volume" (also known as "thin provisioning") can be created by specifying the -s option to the zfs create -V command, or by changing the reservation after the volume has been created. A "sparse volume" is a volume where the reservation is less then the volume size. Consequently, writes to a sparse volume can fail with ENOSPC when the pool is low on space. For a sparse volume, changes to volsize are not reflected in the reservation.
```

In short, the way the option in this project is currently coded doesn't really make a lot of sense (omitting setting the value explicitly leaves it enabled, enabling it sets it to a likely undesired value as well).

I'll probably update the logic to as follows:

```
if zvolEnableReservation === false:
  # add -s to the options used for zfs create
else
  # nothing, let the zfs defaults kick in (ie: auto as the value)
```

**Comment by ishioni** [comment]
Yep. For reference, [this](https://github.com/hpe-storage/truenas-csp/blob/master/truenascsp/truenascsp.py#L311) is how CSP does the post request, you'll note it has the sparse option there explicitly

**Comment by travisghansen** [comment]
Wait, are you using scale with the experimental driver?

**Comment by ishioni** [comment]
I'm using scale, but what experimental driver are you talking about?

**Comment by travisghansen** [comment]
The api-only variant that does not require ssh access? `freenas-api-iscsi`? vs `freenas-iscsi`?

**Comment by ishioni** [comment]
Yep, that' correct. Here's my current setup (i've mangled the files a bit since my last post) - https://github.com/ishioni/k3s-cluster/blob/master/cluster/core/storage/democratic-csi/helm-release.yaml

**Comment by travisghansen** [comment]
OK, that possibly explains the behavior as the TrueNAS api is likely specially handling the `sparse` option vs what I send in explicitly for the `refreservation`. I'll do some more testing and have a fix shortly.

**Comment by travisghansen** [comment]
Fix here: https://github.com/democratic-csi/democratic-csi/commit/4b7604d278a70377f19f21f3d7b512d57ad55aa4

Can you try the `next` image tag (make sure it pulls the most recent if you already used the image) and let me know how it does?

**Comment by ishioni** [comment]
For node, controller or both?

**Comment by travisghansen** [comment]
Only controller should be required.

**Comment by ishioni** [comment]
Looks like I'm hitting something else now

```
code = OutOfRange desc = volume capacity is smaller than the minimum: 1073741824
```

**Comment by travisghansen** [comment]
Oh, I'll fix that. Was thinking the min applied to zvols as well but must just be datasets :( ..one moment.

**Comment by travisghansen** [comment]
Fix is here: https://github.com/democratic-csi/democratic-csi/commit/72856f8da75f4c7bfb055c0f9d612607f60112b8

Will take 20-30 minutes to test/build: https://github.com/democratic-csi/democratic-csi/actions/runs/1518524020

Then give it another try..

**Comment by ishioni** [comment]
Tried it, it now works as expected :)

**Comment by travisghansen** [comment]
I’ll only close this after a new release is snapped with the changes. Thanks for reporting and helping to sort it out!

**Comment by travisghansen** [comment]
Released in `v1.4.3`.
